src/webapi/mutation_observer.rs

```diff
@@ -4,7 +4,60 @@ use webcore::try_from::{TryFrom, TryInto};
 use webapi::node::{INode, Node};
 
 
-/// `MutationObserver` provides a way to receive notifications about changes in the DOM.
+/// The `IMutationObserver` interface represents objects which can receive notifications about changes to the DOM.
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver)
+pub trait IMutationObserver {
```

Comment:
Hmm... the `IMutationObserver` is probably unnecessary. We can just implement `Deref` for `MutationObserverHandle` and have it return `MutationObserver`. (The `MutationObserverHandle` is, after all, only here to handle the lifetime of the observer; it's not a *different* kind of an object implementing the same interface.)

Reply:
I think including `IMutationObserver` makes sense because:

* It's an interface in the DOM spec, which means that theoretically multiple objects can implement it.

   Right now there's only one object implementing it, but making it a `trait` is forwards-compatible.

* It's consistent with other interfaces like `IElement` and `INode`.

But I don't have strong opinions about this, so I don't mind removing `IMutationObserver`

Implementing `Deref` sounds like a good idea, I'll do that.

Reply:
It is indeed an interface, but if only a single object implements it then there isn't really much point - it just adds extra complexity for not much gain. It's not like the interfaces are actually anything concrete in JS either - they're just contracts to which a given type must adhere.

While it's common in some object oriented languages to have useless interfaces which look good on paper I'd like to think that in Rust we tend to not do that if there isn't actually a good practical reason for it.

Reply:
My experience is not with OOP languages, it is with functional languages like Haskell, F#, etc. which use interfaces / typeclasses copiously (and for good reason!).

I can understand why you would think that adding in traits with a single implementation is useless, but there are some practical advantages to it:

* It is easier to write unit tests, because you can simply create a mock object which implements the trait.

* Users can create their own custom implementations, outside of `stdweb`.

   For example, I can imagine React Native (or similar) creating an implementation of `IMutationObserver` for iOS / Android.

   Or a virtual DOM library could implement `IMutationObserver` for virtual DOM nodes.

* If a user writes a function with this type signature...

   ```rust
   fn foo<A: IMutationObserver>(observer: &A)
   ```

   ...then it is forward-compatible and will work with all present and future implementations, but if the user writes this type signature...

   ```rust
   fn foo(observer: &MutationObserver)
   ```

   ...then it can *only* be used with `MutationObserver`, nothing else. This hurts code reuse.

* Traits let you precisely specify your intentions in the code, so the code uses exactly the methods that it needs to use, and no more. Using a concrete type allows it to use *any* method on the concrete type, which makes it harder to understand what the code is doing.

   As an example, if I see `fn foo<A: IMutationObserver>(observer: &A)` then I know *exactly* what methods it can use (`observe`, `disconnect`, and `take_records`), but if I see `fn foo(observer: &MutationObserver)` then it's harder to know which methods it's using or not (without looking at the implementation of `foo` and all the functions that `foo` calls).

* Trait objects let you combine multiple different types together into the same data structure (e.g. a vector), whereas you can't combine concrete types like `Element` and `Node` together (without casting them first).

As such, I don't think using traits is useless, even if they only have a single implementation. In my opinion the complexity of using traits is only slightly more than the complexity of not using traits, and you gain great flexibility by using traits.

I would even go so far as to say that we should actively avoid using concrete types like `Node` and `Element` and should instead prefer trait objects like `&INode`. But that's going pretty far off topic...

In any case, as I said I don't have strong opinions about this, so I'll remove `IMutationObserver`